wiring/inc/spark_wiring_ble.h

```diff
@@ -0,0 +1,460 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, see <http://www.gnu.org/licenses/>.
+ */
+
+#ifndef SPARK_WIRING_BLE_H
+#define SPARK_WIRING_BLE_H
+
+#include <stdint.h>
+#include <stddef.h>
+#include <string.h>
+#include "spark_wiring_platform.h"
+
+#if Wiring_BLE
+
+#include "system_error.h"
+#include "spark_wiring_string.h"
+#include "spark_wiring_vector.h"
+#include "spark_wiring_flags.h"
+#include "ble_hal.h"
+#include "check.h"
+#include <memory>
+
+using spark::Vector;
+using particle::Flags;
+
+class BleScanResult;
+class BlePeerDevice;
+
+// Forward declaration
+class BleCharacteristicImpl;
+class BleServiceImpl;
+class BleGattServerImpl;
+class BleGattClientImpl;
+class BleBroadcasterImpl;
+class BleObserverImpl;
+class BlePeripheralImpl;
+class BleCentralImpl;
+
+enum class BleUuidType {
+    SHORT = 0,
+    LONG = 1
+};
+
+enum class BleUuidOrder {
+    MSB = 0,
+    LSB = 1
+};
+
+struct BleDeviceRoleType; // Tag type for BLE device role flags
+typedef Flags<BleDeviceRoleType, uint8_t> BleDeviceRoles;
+typedef BleDeviceRoles::FlagType BleDeviceRole;
+namespace ROLE {
+    const BleDeviceRole INVALID(BLE_ROLE_INVALID);
+    const BleDeviceRole PERIPHERAL(BLE_ROLE_PERIPHERAL);
+    const BleDeviceRole CENTRAL(BLE_ROLE_CENTRAL);
+    const BleDeviceRole ALL(0xff);
+}
+
+struct BleCharacteristicPropertyType; // Tag type for BLE attribute property flags
+typedef Flags<BleCharacteristicPropertyType, uint8_t> BleCharacteristicProperties;
+typedef BleCharacteristicProperties::FlagType BleCharacteristicProperty;
+namespace PROPERTY {
+    const BleCharacteristicProperty NONE(0);
+    const BleCharacteristicProperty READ(BLE_SIG_CHAR_PROP_READ);
+    const BleCharacteristicProperty WRITE(BLE_SIG_CHAR_PROP_WRITE);
+    const BleCharacteristicProperty WRITE_WO_RSP(BLE_SIG_CHAR_PROP_WRITE_WO_RESP);
+    const BleCharacteristicProperty NOTIFY(BLE_SIG_CHAR_PROP_NOTIFY);
+    const BleCharacteristicProperty INDICATE(BLE_SIG_CHAR_PROP_INDICATE);
+}
+
+namespace PARTICLE_BLE {
+    const uint8_t BLE_USER_DEFAULT_SVC_UUID[BLE_SIG_UUID_128BIT_LEN] = {
+        0x2d, 0x49, 0xc0, 0x26, 0xf0, 0xdb, 0xce, 0x06, 0x7a, 0x33, 0x6f, 0x12, 0x00, 0x00, 0xf3, 0x5d
+    };
+}
+
+typedef uint16_t BleConnectionHandle;
+typedef uint16_t BleAttributeHandle;
+
+typedef void (*OnDataReceivedCallback)(const uint8_t* data, size_t len, const BlePeerDevice& peer, void* context);
+typedef void (*OnScanResultCallback)(const BleScanResult* device);
+typedef void (*OnConnectedCallback)(const BlePeerDevice& peer);
+typedef void (*OnDisconnectedCallback)(const BlePeerDevice& peer);
+
+class BleAddress : public hal_ble_addr_t {
+public:
+    void operator = (hal_ble_addr_t addr) {
```

Comment:
Nitpicking: `operator=(`.
The general practice also would be to return `BleAddress&` (`*this`).